src/Parsers/Kusto/KustoFunctions/IParserKQLFunction.cpp

```diff
@@ -0,0 +1,382 @@
+        return DB::TokenType::ClosingCurlyBrace;
+    else if (token_type == DB::TokenType::OpeningRoundBracket)
+        return DB::TokenType::ClosingRoundBracket;
+    else if (token_type == DB::TokenType::OpeningSquareBracket)
+        return DB::TokenType::ClosingSquareBracket;
+
+    throw DB::Exception(DB::ErrorCodes::NOT_IMPLEMENTED, "Unhandled token: {}", magic_enum::enum_name(token_type));
+}
+
+constexpr bool isClosingBracket(const DB::TokenType token_type)
+{
+    return token_type == DB::TokenType::ClosingCurlyBrace || token_type == DB::TokenType::ClosingRoundBracket
+        || token_type == DB::TokenType::ClosingSquareBracket;
+}
+
+constexpr bool isOpeningBracket(const DB::TokenType token_type)
+{
+    return token_type == DB::TokenType::OpeningCurlyBrace || token_type == DB::TokenType::OpeningRoundBracket
+        || token_type == DB::TokenType::OpeningSquareBracket;
+}
+}
+
+namespace DB
+{
+bool IParserKQLFunction::convert(String & out, IParser::Pos & pos)
+{
+    return wrapConvertImpl(
+        pos,
+        IncreaseDepthTag{},
+        [&]
+        {
+            bool res = convertImpl(out, pos);
+            if (!res)
+                out = "";
+            return res;
+        });
+}
+
+bool IParserKQLFunction::directMapping(
+    String & out, IParser::Pos & pos, const std::string_view ch_fn, const Interval & argument_count_interval)
+{
+    const auto fn_name = getKQLFunctionName(pos);
+    if (fn_name.empty())
+        return false;
+
+    out.append(ch_fn.data(), ch_fn.length());
+    out.push_back('(');
+
+    int argument_count = 0;
+    const auto begin = pos;
+    while (!pos->isEnd() && pos->type != TokenType::PipeMark && pos->type != TokenType::Semicolon)
+    {
+        if (pos != begin)
+            out.append(", ");
+
+        if (const auto argument = getOptionalArgument(fn_name, pos))
+        {
+            ++argument_count;
+            out.append(*argument);
+        }
+
+        if (pos->type == TokenType::ClosingRoundBracket)
+        {
+            if (!argument_count_interval.IsWithinBounds(argument_count))
+                throw Exception(
+                    ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH,
+                    "{}: between {} and {} arguments are expected, but {} were provided",
+                    fn_name,
+                    argument_count_interval.Min(),
+                    argument_count_interval.Max(),
+                    argument_count);
+
+            out.push_back(')');
+            return true;
+        }
+    }
+
+    out.clear();
+    pos = begin;
+    return false;
+}
+
+String IParserKQLFunction::generateUniqueIdentifier()
+{
+    // This particular random generator hits each number exactly once before looping over.
+    // Because of this, it's sufficient for queries consisting of up to 2^16 (= 65536) distinct function calls.
+    // Reference: https://www.pcg-random.org/using-pcg-cpp.html#insecure-generators
+    static pcg32_once_insecure random_generator;
```

Comment:
This generator is not thread safe - https://s3.amazonaws.com/clickhouse-test-reports/55418/769ed2e19d46fcb9cb6a678a0da6d6f2fc5d239e/stateless_tests__tsan__[5_5]/stderr.log

Reply:
https://github.com/ClickHouse/ClickHouse/pull/55626

Reply:
yes, we know - already fixed